analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnModulus.cs

```diff
@@ -50,27 +61,62 @@ private static void VisitEquality(SyntaxNodeAnalysisContext c)
             if (CheckExpression(equalsExpression.Left, equalsExpression.Right, c.SemanticModel, out var constantValue)
                 || CheckExpression(equalsExpression.Right, equalsExpression.Left, c.SemanticModel, out constantValue))
             {
-                c.ReportIssue(Diagnostic.Create(Rule,
-                                                               equalsExpression.GetLocation(),
-                                                               constantValue < 0 ? "negative" : "positive"));
+                c.ReportIssue(Diagnostic.Create(Rule, equalsExpression.GetLocation(), constantValue < 0 ? "negative" : "positive"));
             }
         }
 
         private static bool CheckExpression(SyntaxNode node, ExpressionSyntax modulus, SemanticModel semanticModel, out int constantValue) =>
             ExpressionNumericConverter.TryGetConstantIntValue(node, out constantValue)
             && constantValue != 0
             && IsModulus(modulus)
-            && !IsUnsigned(modulus, semanticModel);
+            && !ExpressionIsAlwaysPositive(modulus, semanticModel);
 
         private static bool IsModulus(ExpressionSyntax expression) =>
             expression.RemoveParentheses() is BinaryExpressionSyntax binary
             && binary.IsKind(SyntaxKind.ModuloExpression);
 
-        private static bool IsUnsigned(ExpressionSyntax expression, SemanticModel semantic)
+        private static bool ExpressionIsAlwaysPositive(ExpressionSyntax expression, SemanticModel semantic)
         {
             var type = semantic.GetTypeInfo(expression).Type;
-            return type.IsAny(KnownType.UnsignedIntegers)
-                   || type.Is(KnownType.System_UIntPtr);
+            var isUint = type.IsAny(KnownType.UnsignedIntegers)
+                          || type.Is(KnownType.System_UIntPtr);
+
+            var leftExpression = ((BinaryExpressionSyntax)expression).Left;
+            if (!isUint && CollectionSizePropertyOrMethodNames.Any(x => leftExpression.ToString().Contains(x)))
+            {
+                var symbol = semantic.GetSymbolInfo(((BinaryExpressionSyntax)expression).Left).Symbol;
+                return IsCollectionSizeMethodOrProperty(symbol);
+            }
+            return isUint;
```

Comment:
this can then be
```suggestion
            return CollectionSizePropertyOrMethodNames.Any(x => leftExpression.ToString().Contains(x)))
              && semantic.GetSymbolInfo(((BinaryExpressionSyntax)expression).Left).Symbol is { } symbol
              && IsCollectionSizeMethodOrProperty(symbol);

```